### PR TITLE
Icon registry for kirby icons

### PR DESCRIPTION
**Pull request by jvh2**
This PR adds functionality for kirby icons, so that provided icons dont override each other, but is still backwards compatible with old solution in DRB.

Current way of adding custom icons in @NGModule
```
providers: [
 { 
    provide: ICON_SETTINGS, 
    useValue: iconSettings 
 }
]
```
New way of adding custom icons
```
export class @NGModule {
  constructor(kirbyIconRegistryService: KirbyIconRegistryService) {
    kirbyIconRegistryService.addIcons(this.icons);
  }
}
```
namespace for icons keys is not part of this PR

### CODE REVIEW

File: libs/designsystem/src/lib/components/icon/icon.component.spec.ts [hunk 1]
```diff
@@ -113,6 +115,24 @@ describe('IconComponent', () => {
...
+  describe('multiple icon settings', () => {
+    it('should combine multiple provided icon, but use icons from ICON_SETTINGS', () => {
+      const fixture = createTestComponent(`<kirby-icon customName="name1"></kirby-icon>`);
+      fixture.detectChanges();
+      const component = fixture.debugElement.query(By.directive(IconComponent)).componentInstance;
+      expect(component.icon.name).toBe('name1');
+      expect(component.icon.svg).toBe('svg1');
+    });
+
+    it('should combine multiple provided icon, and use icons added with KirbyIconRegistryService', () => {
```
Comment (lydemann) [inline libs/designsystem/src/lib/components/icon/icon.component.spec.ts:127]:
```suggestion
    it('should combine multiple provided icons, and use icons added with KirbyIconRegistryService', () => {
```

File: libs/designsystem/src/lib/components/icon/icon.component.spec.ts [hunk 2]
```diff
@@ -113,6 +115,24 @@ describe('IconComponent', () => {
       });
     });
   });
+  describe('multiple icon settings', () => {
+    it('should combine multiple provided icon, but use icons from ICON_SETTINGS', () => {
```
Comment (lydemann) [inline libs/designsystem/src/lib/components/icon/icon.component.spec.ts:119]:
```suggestion
    it('should combine multiple provided icons, but use icons from ICON_SETTINGS', () => {
```

File: libs/designsystem/src/lib/components/icon/icon.component.ts [hunk 3]
```diff
@@ -54,13 +57,33 @@ export class IconComponent implements OnChanges {
...
+  constructor(
+    private iconRegistryService: KirbyIconRegistryService,
+    @Optional() @Inject(ICON_SETTINGS) private iconSettings?: IconSettings
+  ) {
+    this.combineIconSettings();
+  }
+
+  private combineIconSettings() {
+    const iconSettings: Icon[] =
+      this.iconSettings && this.iconSettings.icons
```
Comment (lydemann) [inline libs/designsystem/src/lib/components/icon/icon.component.ts:73]: You can just:

`
let iconSettingsIcons = this.iconSettings?.icons;
if(!icons) {
iconSettingsIcons = [this.defaultIcon]
}
`

File: libs/designsystem/src/lib/components/icon/icon.component.ts [hunk 4]
```diff
@@ -54,13 +57,33 @@ export class IconComponent implements OnChanges {
...
+   */
+  constructor(
+    private iconRegistryService: KirbyIconRegistryService,
+    @Optional() @Inject(ICON_SETTINGS) private iconSettings?: IconSettings
+  ) {
+    this.combineIconSettings();
+  }
+
+  private combineIconSettings() {
+    const iconSettings: Icon[] =
```
Comment (lydemann) [inline libs/designsystem/src/lib/components/icon/icon.component.ts:72]: This is not iconSettings but icons from icon settings

File: package.json [hunk 5]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "designsystem",
-  "version": "1.2.77",
+  "version": "1.2.80",
```
Comment (lydemann) [inline package.json:3]: Remember to update verison in package.json of designsystem as well.

File: libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
+import { Injectable } from '@angular/core';
+
+import { Icon } from './icon-settings';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class KirbyIconRegistryService {
+  private iconRegistry = new Map<string, string>();
+  constructor() {}
```
Comment (lydemann) [inline libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.ts:10]:
```suggestion
```

File: libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+  addIcon(iconName: string, svgPath: string): void {
+    this.iconRegistry.set(iconName, svgPath);
+  }
+
+  addIcons(icons: Icon[]): void {
+    if (!icons) {
+      console.error('Icon array not defined');
```
Comment (lydemann) [inline libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.ts:18]:
```suggestion
      console.error('Icons not defined');
```

**Review by lydemann — CHANGES_REQUESTED** [review]
Some remarks

File: libs/designsystem/src/lib/components/icon/icon.component.ts (hunk 4 above)
Comment (jvh2) [reply libs/designsystem/src/lib/components/icon/icon.component.ts:72]: 👍

File: libs/designsystem/src/lib/components/icon/icon.component.ts (hunk 3 above)
Comment (jakobe) [reply libs/designsystem/src/lib/components/icon/icon.component.ts:73]: @lydemann Nope, we are on TypeScript 3.6.5 for compability reasons

File: package.json (hunk 5 above)
Comment (jakobe) [reply package.json]: You shouldn't bump the version number yourself unless you are making major/minor changes

**Review by lydemann — APPROVED** [review]

File: libs/designsystem/src/lib/components/icon/icon.component.ts [hunk 8]
```diff
@@ -54,13 +57,33 @@ export class IconComponent implements OnChanges {
...
+
+  private combineIconSettings() {
+    const iconSettingsIcons: Icon[] =
+      this.iconSettings && this.iconSettings.icons
+        ? [...this.iconSettings.icons]
+        : [this.defaultIcon];
+
+    this.combinedIconSettings = {
+      icons: [...iconSettingsIcons, ...this.iconRegistryService.getCustomIcons()],
+    } as IconSettings;
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/icon/icon.component.ts:79]: Consider simplyfing:
```suggestion
    const iconSettingsIcons = this.iconSettings ? this.iconSettings.icons : [];
    this.registeredIcons = [...iconSettingsIcons, ...this.iconRegistryService.getCustomIcons()];
```

File: libs/designsystem/src/lib/components/icon/icon.component.ts [hunk 9]
```diff
@@ -54,13 +57,33 @@ export class IconComponent implements OnChanges {
...
+  }
 
   ngOnChanges(changes: SimpleChanges): void {
     if (changes.name && changes.name.currentValue) {
       this.icon = this.findIcon(kirbyIconSettings.icons, changes.name.currentValue);
     } else if (changes.customName && changes.customName.currentValue) {
-      this.icon = this.findIcon(this.iconSettings.icons, changes.customName.currentValue);
+      this.icon = this.findIcon(this.combinedIconSettings.icons, changes.customName.currentValue);
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/icon/icon.component.ts:86]:
```suggestion
      this.icon = this.findIcon(this.registeredIcons, changes.customName.currentValue);
```

File: libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.spec.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+  let service: KirbyIconRegistryService;
+
+  beforeEach(() => {
+    service = new KirbyIconRegistryService();
+  });
+  it('should create service', () => {
+    expect(service).toBeTruthy();
+  });
+  describe('getCustomIcons', () => {
+    it('get icons from empty map', () => {
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.spec.ts:14]:
```suggestion
    it('should return empty array by default', () => {
```

File: libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.spec.ts [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+      const expectedIcons: Icon[] = [];
+      expect(service.getCustomIcons()).toEqual(expectedIcons);
+    });
+    it('get excisting icons in map', () => {
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.spec.ts:18]:
```suggestion
    it('should return registered icons', () => {
```

File: libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.spec.ts [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+  beforeEach(() => {
+    service = new KirbyIconRegistryService();
+  });
+  it('should create service', () => {
+    expect(service).toBeTruthy();
+  });
+  describe('getCustomIcons', () => {
+    it('get icons from empty map', () => {
+      const expectedIcons: Icon[] = [];
+      expect(service.getCustomIcons()).toEqual(expectedIcons);
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.spec.ts:16]:
```suggestion
      const emptyArray = [];
      expect(service.getCustomIcons()).toEqual(emptyArray);
```

File: libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.spec.ts [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+        icons: [
+          { name: 'name1', svg: 'svg1' },
+          { name: 'name2', svg: 'svg2' },
+        ] as Icon[],
+      } as IconSettings;
+      service.addIcons(IconSettings.icons);
+      const expectedIcons = [
+        { name: 'name1', svg: 'svg1' },
+        { name: 'name2', svg: 'svg2' },
+      ];
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.spec.ts:29]: Consider simplyfing:
```suggestion
      const expectedIcons = [
        { name: 'name1', svg: 'svg1' },
        { name: 'name2', svg: 'svg2' },
      ];
      service.addIcons(expectedIcons);
```

File: libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.ts [hunk 14]
```diff
@@ -0,0 +1,30 @@
+import { Injectable } from '@angular/core';
+
+import { Icon } from './icon-settings';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class KirbyIconRegistryService {
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.ts:8]: Consider naming - we don't use the `Kirby` prefix in general, as it's already exported to / imported from `@kirbydesign/designsystem`
```suggestion
export class IconRegistryService {
```

File: libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.ts [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+    if (!icons) {
+      console.error('Icons not defined');
+      return;
+    }
+    icons.forEach((icon) => {
+      this.addIcon(icon.name, icon.svg);
+    });
+  }
+
+  getCustomIcons(): Icon[] {
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.ts:25]: Consider naming:
```suggestion
  getIcons(): Icon[] {
```

File: package.json [hunk 16]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "designsystem",
-  "version": "1.2.81",
+  "version": "1.2.82",
```
Comment (jakobe) [inline package.json:3]: Don't bump this manually...

**Review by jakobe — CHANGES_REQUESTED** [review]

File: libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.ts (hunk 14 above)
Comment (lydemann) [reply libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.ts:8]: Might be beneficial here. Look how Angular Material prefixes their registry service with eg. MatIconRegistry.

https://material.angular.io/components/icon/api

File: libs/designsystem/src/lib/components/icon/icon.component.spec.ts [hunk 17]
```diff
@@ -113,6 +115,24 @@ describe('IconComponent', () => {
       });
     });
   });
+  describe('multiple icon settings', () => {
+    it('should combine multiple provided icons, but use icons from ICON_SETTINGS', () => {
```
[the remark this answers is not in the document]
Comment (jvh2) [reply libs/designsystem/src/lib/components/icon/icon.component.spec.ts:119]: Test moved to icon describe section, and naming changed. Purpose of test is to verify both ICON_SETTINGS and KirbyIconRegistryService works with custom icons. Test validates that combined icons works for both provided settings and icons added with KirbyIconRegistryService

File: libs/designsystem/src/lib/components/icon/icon.component.spec.ts [hunk 18]
```diff
@@ -113,6 +115,24 @@ describe('IconComponent', () => {
...
+      const fixture = createTestComponent(`<kirby-icon customName="name1"></kirby-icon>`);
+      fixture.detectChanges();
+      const component = fixture.debugElement.query(By.directive(IconComponent)).componentInstance;
+      expect(component.icon.name).toBe('name1');
+      expect(component.icon.svg).toBe('svg1');
+    });
+
+    it('should combine multiple provided icons, and use icons added with KirbyIconRegistryService', () => {
```
[the remark this answers is not in the document]
Comment (jvh2) [reply libs/designsystem/src/lib/components/icon/icon.component.spec.ts:127]: se above

File: libs/designsystem/src/lib/components/icon/icon.component.ts [hunk 19]
```diff
@@ -54,13 +57,33 @@ export class IconComponent implements OnChanges {
...
+    @Optional() @Inject(ICON_SETTINGS) private iconSettings?: IconSettings
+  ) {
+    this.combineIconSettings();
+  }
+
+  private combineIconSettings() {
+    const iconSettingsIcons: Icon[] =
+      this.iconSettings && this.iconSettings.icons
+        ? [...this.iconSettings.icons]
+        : [this.defaultIcon];
```
[the remark this answers is not in the document]
Comment (jvh2) [reply libs/designsystem/src/lib/components/icon/icon.component.ts:75]: 👍

File: libs/designsystem/src/lib/components/icon/icon.component.ts (hunk 8 above)
Comment (jvh2) [reply libs/designsystem/src/lib/components/icon/icon.component.ts:79]: 👍

File: libs/designsystem/src/lib/components/icon/icon.component.ts (hunk 9 above)
Comment (jvh2) [reply libs/designsystem/src/lib/components/icon/icon.component.ts:86]: For nu synes jeg combinedIconSettings giver bedre mening da det jo er en "combined" IconSettings. Den håndtere både provided icons samt icons added med KirbyIconRegistryService

Når ICON_SETTINGS er helt deprecated kan det refactoreret

File: libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.spec.ts (hunk 12 above)
Comment (jvh2) [reply libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.spec.ts:16]: changed to const expectedIcons = [];
Dont like datatypes in varible name :-)

File: libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.spec.ts (hunk 13 above)
Comment (jvh2) [reply libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.spec.ts:29]: 👍

File: libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.ts (hunk 14 above)
Comment (jvh2) [reply libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.ts:8]: An also its more explicit that its a kirby service, without having to look at the package import.

File: package.json (hunk 16 above)
Comment (jvh2) [reply package.json:3]: OK :-)

File: libs/designsystem/src/lib/components/icon/icon.component.spec.ts [hunk 20]
```diff
@@ -47,6 +47,22 @@ describe('IconComponent', () => {
       expect(component.name).toBe(undefined);
       expect(component.icon.name).toBe(component.defaultIcon.name);
     });
+
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/icon/icon.component.spec.ts:50]: Consider adding missing sunshine test; `it('should use default icons from Kirby icon settings'` ...

File: libs/designsystem/src/lib/components/icon/icon.component.spec.ts [hunk 21]
```diff
@@ -47,6 +47,22 @@ describe('IconComponent', () => {
...
+    it('should use custom icons from ICON_SETTINGS', () => {
+      const fixture = createTestComponent(`<kirby-icon customName="name1"></kirby-icon>`);
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/icon/icon.component.spec.ts:52]: Consider icon naming (to reflect the icon comes from a custom ICON_SETTINGS)
```suggestion
      const fixture = createTestComponent(`<kirby-icon customName="customIconFromIconSettings"></kirby-icon>`);
```

File: libs/designsystem/src/lib/components/icon/icon.component.spec.ts [hunk 22]
```diff
@@ -47,6 +47,22 @@ describe('IconComponent', () => {
...
+      fixture.detectChanges();
+      const component = fixture.debugElement.query(By.directive(IconComponent)).componentInstance;
+      expect(component.icon.name).toBe('name1');
+      expect(component.icon.svg).toBe('svg1');
+    });
+
+    it('should use custom icons added with KirbyIconRegistryService', () => {
+      const fixture = createTestComponent(`<kirby-icon customName="customName1"></kirby-icon>`);
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/icon/icon.component.spec.ts:60]: Consider icon naming (to reflect the icon comes from a custom ICON_SETTINGS)
```suggestion
      const fixture = createTestComponent(`<kirby-icon customName="customIconFromIconRegistry"></kirby-icon>`);
```

File: libs/designsystem/src/lib/components/icon/icon.component.spec.ts (hunk 20 above)
Comment (jakobe) [inline libs/designsystem/src/lib/components/icon/icon.component.spec.ts:50]: Consider adding test that describes what happens if icon with same name is added BOTH through IconSettings AND Registry...

File: libs/designsystem/src/lib/components/icon/icon.component.spec.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (jakobe) [reply libs/designsystem/src/lib/components/icon/icon.component.spec.ts]: Ok check - then consider adding test that describes what happens if icon with same name is added BOTH through IconSettings AND Registry...

File: libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.ts (hunk 14 above)
Comment (jakobe) [reply libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.ts:8]: That might be, but consistency is of importance here, so please keep naming consistent :)

File: libs/designsystem/src/lib/components/icon/icon.component.ts [hunk 23]
```diff
@@ -69,14 +69,10 @@ export class IconComponent implements OnChanges {
   }
 
   private combineIconSettings() {
-    const iconSettingsIcons: Icon[] =
-      this.iconSettings && this.iconSettings.icons
-        ? [...this.iconSettings.icons]
-        : [this.defaultIcon];
-
+    const iconSettingsIcons = this.iconSettings ? this.iconSettings.icons : [];
     this.combinedIconSettings = {
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/icon/icon.component.ts:73]: Consider renaming
```suggestion
    this.icons = {
```

File: libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.spec.ts (hunk 12 above)
Comment (jakobe) [reply libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.spec.ts]: Ok - but it's just not very clear that what you are REALLY expecting is that icons should be empty...
Something a long the lines of `expect(service.getCustomIcons()).toBeEmpty()` (if jasmine had that syntax)

**Review by jakobe — CHANGES_REQUESTED** [review]
See comments

File: libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.ts (hunk 14 above)
Comment (jvh2) [reply libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.ts:8]: Valid point if kirby prefix not used elsewhere!

File: libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.ts (hunk 15 above)
Comment (jvh2) [reply libs/designsystem/src/lib/components/icon/kirby-icon-registry.service.ts:25]: 👍

File: libs/designsystem/src/lib/components/icon/icon.component.spec.ts (hunk 22 above)
Comment (jvh2) [reply libs/designsystem/src/lib/components/icon/icon.component.spec.ts:60]: 👍

File: libs/designsystem/src/lib/components/icon/icon.component.spec.ts (hunk 20 above)
Comment (jvh2) [reply libs/designsystem/src/lib/components/icon/icon.component.spec.ts:50]: Impl. as we talked about. IconSettings and Registry icons is merged together with warn on overwrite

File: libs/designsystem/src/lib/components/icon/icon.component.ts (hunk 23 above)
Comment (jvh2) [reply libs/designsystem/src/lib/components/icon/icon.component.ts:73]: combinedIconSettings removed

File: libs/designsystem/src/lib/components/icon/icon.component.spec.ts (hunk 20 above)
Comment (jvh2) [reply libs/designsystem/src/lib/components/icon/icon.component.spec.ts:50]: added

File: libs/designsystem/src/lib/components/icon/icon.component.spec.ts (hunk 21 above)
Comment (jvh2) [reply libs/designsystem/src/lib/components/icon/icon.component.spec.ts:52]: 👍

File: libs/designsystem/src/lib/components/icon/icon.component.spec.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (jvh2) [reply libs/designsystem/src/lib/components/icon/icon.component.spec.ts]: as we talked about they are merged together with warning on key collision

**Review by jakobe — APPROVED** [review]
giphub

**Pull request merged**